From: Vlad Buslov <vladbu@mellanox.com>
Date: Mon, 24 Sep 2018 19:22:49 +0300
Subject: net: core: netlink: add helper refcount dec and lock function
Patch-mainline: v4.20-rc1
Git-commit: 6f99528e9797794b91b43321fbbc93fe772b0803
References: bsc#1196973 CVE-2021-39713

Rtnl lock is encapsulated in netlink and cannot be accessed by other
modules directly. This means that reference counted objects that rely on
rtnl lock cannot use it with refcounter helper function that atomically
releases decrements reference and obtains mutex.

This patch implements simple wrapper function around refcount_dec_and_lock
that obtains rtnl lock if reference counter value reached 0.

SUSE: Changes refcount_* to atomic_* as older codestreams are not
      converted to refcount API.

Signed-off-by: Vlad Buslov <vladbu@mellanox.com>
Acked-by: Jiri Pirko <jiri@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 include/linux/rtnetlink.h |    1 +
 net/core/rtnetlink.c      |    6 ++++++
 2 files changed, 7 insertions(+)

--- a/include/linux/rtnetlink.h
+++ b/include/linux/rtnetlink.h
@@ -29,6 +29,7 @@ extern void rtnl_lock(void);
 extern void rtnl_unlock(void);
 extern int rtnl_trylock(void);
 extern int rtnl_is_locked(void);
+extern bool atomic_dec_and_rtnl_lock(atomic_t *r);
 
 extern wait_queue_head_t netdev_unregistering_wq;
 extern struct mutex net_mutex;
--- a/net/core/rtnetlink.c
+++ b/net/core/rtnetlink.c
@@ -119,6 +119,12 @@ int rtnl_is_locked(void)
 }
 EXPORT_SYMBOL(rtnl_is_locked);
 
+bool atomic_dec_and_rtnl_lock(atomic_t *r)
+{
+	return atomic_dec_and_mutex_lock(r, &rtnl_mutex);
+}
+EXPORT_SYMBOL(atomic_dec_and_rtnl_lock);
+
 #ifdef CONFIG_PROVE_LOCKING
 bool lockdep_rtnl_is_held(void)
 {
